gum/backend-windows/gumprocess-windows.c

```diff
@@ -305,13 +305,77 @@ _gum_process_enumerate_ranges (GumPageProtection prot,
   }
 }
 
+static gboolean
+gum_process_enumerate_heap_ranges (HANDLE heap,
+                                   GumFoundMallocRangeFunc func,
+                                   gpointer user_data)
+{
+  gboolean carry_on;
+  gboolean locked_heap;
+  GumMemoryRange range;
+  GumMallocRangeDetails details;
+  PROCESS_HEAP_ENTRY entry;
+
+  /* HeapLock may fail but it doesn't seem to have any real consequences... */
+  locked_heap = HeapLock (heap);
+
+  details.range = &range;
+  carry_on = TRUE;
+  entry.lpData = NULL;
+  while (carry_on && HeapWalk (heap, &entry))
+  {
+    if ((entry.wFlags & PROCESS_HEAP_ENTRY_BUSY) != 0)
+    {
+      range.base_address = (GumAddress) entry.lpData;
+      range.size = entry.cbData;
+      carry_on = func (&details, user_data);
+    }
+  }
+
+  if (locked_heap)
+    HeapUnlock (heap);
+
+  return carry_on;
+}
+
 void
 gum_process_enumerate_malloc_ranges (GumFoundMallocRangeFunc func,
                                      gpointer user_data)
 {
-  /* Not implemented */
-  (void) func;
-  (void) user_data;
+  HANDLE process_heap;
+  DWORD num_heaps;
+  HANDLE * heaps;
+  DWORD i;
+
+  process_heap = GetProcessHeap ();
+  if (!gum_process_enumerate_heap_ranges (process_heap, func, user_data))
+    return;
+
+  num_heaps = GetProcessHeaps (0, NULL);
+  if (num_heaps == 0)
+    return;
+  heaps = HeapAlloc (process_heap, 0, num_heaps * sizeof (HANDLE));
+  if (heaps == NULL)
+    return;
+  num_heaps = GetProcessHeaps (num_heaps, heaps);
+  if (num_heaps == 0)
```

Comment:
Let's simplify this:

```
  if (num_heaps == 0)
    goto beach;
```

and just add a `beach:` label right above the `HeapFree` below.